zkevm-circuits/src/evm_circuit/execution/sdiv_smod.rs

```diff
@@ -0,0 +1,350 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        step::ExecutionState,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition::Delta},
+            math_gadget::{
+                generate_lagrange_base_polynomial, AbsWordGadget, IsZeroGadget, LtGadget,
+                LtWordGadget, MulAddWordsGadget,
+            },
+            select, sum, CachedRegion,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+use bus_mapping::evm::OpcodeId;
+use eth_types::{Field, ToLittleEndian, U256};
+use halo2_proofs::plonk::Error;
+
+#[derive(Clone, Debug)]
+pub(crate) struct SignedDivModGadget<F> {
+    same_context: SameContextGadget<F>,
+    quotient_abs_word: AbsWordGadget<F>,
+    divisor_abs_word: AbsWordGadget<F>,
+    remainder_abs_word: AbsWordGadget<F>,
+    dividend_abs_word: AbsWordGadget<F>,
+    mul_add_words: MulAddWordsGadget<F>,
+    remainder_abs_lt_divisor_abs: LtWordGadget<F>,
+    dividend_is_signed_overflow: LtGadget<F, 1>,
+    quotient_is_zero: IsZeroGadget<F>,
+    divisor_is_zero: IsZeroGadget<F>,
+    remainder_is_zero: IsZeroGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for SignedDivModGadget<F> {
+    const NAME: &'static str = "SDIV_SMOD";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::SDIV_SMOD;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+        let is_sdiv = generate_lagrange_base_polynomial(
```

Comment:
It seems that should not add another gadget.
I fixed to a simple expression without `generate_lagrange_base_polynomial`. Since it is clear for only two opcodes.
```
let is_sdiv = (OpcodeId::SMOD.expr() - opcode.expr()) * F::from(2).invert().unwrap();
```